File: backend/internal/geofence/geofence.go
// Package geofence reports whether a coordinate lies within Chicago
package geofence

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"math"

	"github.com/camden-git/rev-monorepo/backend/internal/h3util"
)

// chicagoGeoJSON is the City of Chicago official boundary simplified to ~33m.
// it is a MultiPolygon
//
//go:embed chicago.geojson
var chicagoGeoJSON []byte

// bufferMeters expands the boundary outward so claims that drift just outside
// the city (e.g. GPS noise along the lakefront) still count as inside
const bufferMeters = 150

// metersPerDegLat is constant; longitude degrees shrink with latitude
const metersPerDegLat = 111_320.0

func metersPerDegLng(lat float64) float64 {
	return metersPerDegLat * math.Cos(lat*math.Pi/180)
}

// ring is a closed list of [lng, lat] vertices, matching GeoJSON axis order
type ring [][2]float64

// polygon is an outer ring followed by zero or more hole rings
type polygon []ring

var chicago []polygon

// bb is Chicago's bounding box, used to reject far-away points before the
// per-vertex ray cast
var bbMinLng, bbMinLat, bbMaxLng, bbMaxLat float64

func init() {
	var geo struct {
		Coordinates []polygon `json:"coordinates"`
	}
	if err := json.Unmarshal(chicagoGeoJSON, &geo); err != nil {
		panic(fmt.Sprintf("geofence: parse chicago boundary: %v", err))
	}
	chicago = geo.Coordinates
	if len(chicago) == 0 {
		panic("geofence: chicago boundary is empty")
	}

	bbMinLng, bbMinLat = math.Inf(1), math.Inf(1)
	bbMaxLng, bbMaxLat = math.Inf(-1), math.Inf(-1)
	for _, poly := range chicago {
		for _, v := range poly[0] {
			bbMinLng, bbMaxLng = math.Min(bbMinLng, v[0]), math.Max(bbMaxLng, v[0])
			bbMinLat, bbMaxLat = math.Min(bbMinLat, v[1]), math.Max(bbMaxLat, v[1])
		}
	}
}

// ContainsLatLng reports whether the coordinate falls inside the Chicago
// boundary, or within bufferMeters of it. a point inside an outer ring but
// within one of its holes is outside
func ContainsLatLng(lat, lng float64) bool {
	bufLat := bufferMeters / metersPerDegLat
	bufLng := bufferMeters / metersPerDegLng(lat)
	if lng < bbMinLng-bufLng || lng > bbMaxLng+bufLng || lat < bbMinLat-bufLat || lat > bbMaxLat+bufLat {
		return false
	}
	for _, poly := range chicago {
		if len(poly) == 0 || !poly[0].contains(lng, lat) {
			continue
		}
		inHole := false
		for _, hole := range poly[1:] {
			if hole.contains(lng, lat) {
				inHole = true
				break
			}
		}
		if !inHole {
			return true
		}
	}
	// outside the polygon, but keep it if it sits within the buffer of any edge
	return withinBuffer(lat, lng)
}

// withinBuffer reports whether the point lies within bufferMeters of any
// boundary edge (outer rings and holes alike)
func withinBuffer(lat, lng float64) bool {
	mLng := metersPerDegLng(lat)
	for _, poly := range chicago {
		for _, r := range poly {
			for i, j := 0, len(r)-1; i < len(r); j, i = i, i+1 {
				if segDistMeters(lng, lat, r[j], r[i], mLng) <= bufferMeters {
					return true
				}
			}
		}
	}
	return false
}

// segDistMeters is the distance from point (plng, plat) to segment a-b, using a
// local equirectangular projection (meters relative to the point)
func segDistMeters(plng, plat float64, a, b [2]float64, mLng float64) float64 {
	ax, ay := (a[0]-plng)*mLng, (a[1]-plat)*metersPerDegLat
	bx, by := (b[0]-plng)*mLng, (b[1]-plat)*metersPerDegLat
	dx, dy := bx-ax, by-ay
	if dx == 0 && dy == 0 {
		return math.Hypot(ax, ay)
	}
	t := -(ax*dx + ay*dy) / (dx*dx + dy*dy)
	t = math.Max(0, math.Min(1, t))
	return math.Hypot(ax+t*dx, ay+t*dy)
}

// ContainsCell reports whether the center of an h3 cell falls inside Chicago
func ContainsCell(id uint64) bool {
	lat, lng, ok := h3util.CellCenter(id)
	if !ok {
		return false
	}
	return ContainsLatLng(lat, lng)
}

// contains runs a ray-casting point-in-polygon test. x is lng, y is lat to
// match the stored [lng, lat] vertex order
func (r ring) contains(x, y float64) bool {
	inside := false
	for i, j := 0, len(r)-1; i < len(r); j, i = i, i+1 {
		xi, yi := r[i][0], r[i][1]
		xj, yj := r[j][0], r[j][1]
		if (yi > y) != (yj > y) && x < (xj-xi)*(y-yi)/(yj-yi)+xi {
			inside = !inside
		}
	}
	return inside
}

File: ios/RevKit/Sources/RevKit/Game/Geofence.swift
import Foundation

/// reports whether a coordinate lies within Chicago, mirroring the server's
/// `internal/geofence` package (same embedded boundary, same 150 m buffer).
///
/// the server silently skips claims outside the city, so without this mirror an
/// out-of-town drive would persist optimistic claims the server never accepts
public enum Geofence {
    /// expands the boundary outward so claims that drift just outside the city
    /// (e.g. GPS noise along the lakefront) still count as inside
    static let bufferMeters = 150.0

    static let metersPerDegLat = 111_320.0

    static func metersPerDegLng(_ lat: Double) -> Double {
        metersPerDegLat * cos(lat * .pi / 180)
    }

    /// a closed list of [lng, lat] vertices (GeoJSON axis order); a polygon is
    /// an outer ring followed by zero or more hole rings
    private typealias Ring = [[Double]]
    private typealias Polygon = [Ring]

    private struct Boundary {
        let polygons: [Polygon]
        let minLng, minLat, maxLng, maxLat: Double
    }

    /// nil when the bundled boundary is missing or unparseable; every check then
    /// fails closed (no claims), matching the server
    private static let boundary: Boundary? = loadBoundary()

    private static func loadBoundary() -> Boundary? {
        struct Geometry: Decodable { let coordinates: [[[[Double]]]] }
        guard let url = Bundle.module.url(forResource: "chicago", withExtension: "geojson"),
              let data = try? Data(contentsOf: url),
              let geo = try? JSONDecoder().decode(Geometry.self, from: data),
              !geo.coordinates.isEmpty
        else {
            assertionFailure("Geofence: bundled chicago.geojson missing or unparseable")
            return nil
        }
        var minLng = Double.infinity, minLat = Double.infinity
        var maxLng = -Double.infinity, maxLat = -Double.infinity
        for poly in geo.coordinates {
            guard let outer = poly.first else { continue }
            for v in outer where v.count >= 2 {
                minLng = Swift.min(minLng, v[0]); maxLng = Swift.max(maxLng, v[0])
                minLat = Swift.min(minLat, v[1]); maxLat = Swift.max(maxLat, v[1])
            }
        }
        return Boundary(polygons: geo.coordinates, minLng: minLng, minLat: minLat, maxLng: maxLng, maxLat: maxLat)
    }

    /// whether the coordinate falls inside the Chicago boundary, or within
    /// `bufferMeters` of it. a point inside an outer ring but within one of its
    /// holes is outside
    public static func containsLatLng(lat: Double, lng: Double) -> Bool {
        guard let boundary else { return false }
        let bufLat = bufferMeters / metersPerDegLat
        let bufLng = bufferMeters / metersPerDegLng(lat)
        if lng < boundary.minLng - bufLng || lng > boundary.maxLng + bufLng
            || lat < boundary.minLat - bufLat || lat > boundary.maxLat + bufLat {
            return false
        }
        for poly in boundary.polygons {
            guard let outer = poly.first, contains(ring: outer, x: lng, y: lat) else { continue }
            let inHole = poly.dropFirst().contains { contains(ring: $0, x: lng, y: lat) }
            if !inHole { return true }
        }
        // outside the polygon, but keep it if it sits within the buffer of any edge
        return withinBuffer(lat: lat, lng: lng)
    }

    /// whether the center of an H3 cell falls inside Chicago
    public static func containsCell(_ id: UInt64) -> Bool {
        guard let center = H3Grid.cellCenter(of: id) else { return false }
        return containsLatLng(lat: center.latitude, lng: center.longitude)
    }

    /// whether the point lies within `bufferMeters` of any boundary edge (outer
    /// rings and holes alike)
    private static func withinBuffer(lat: Double, lng: Double) -> Bool {
        guard let boundary else { return false }
        let mLng = metersPerDegLng(lat)
        for poly in boundary.polygons {
            for ring in poly {
                var j = ring.count - 1
                for i in 0..<ring.count {
                    if segDistMeters(plng: lng, plat: lat, a: ring[j], b: ring[i], mLng: mLng) <= bufferMeters {
                        return true
                    }
                    j = i
                }
            }
        }
        return false
    }

    /// distance from point (plng, plat) to segment a-b, using a local
    /// equirectangular projection (meters relative to the point)
    private static func segDistMeters(plng: Double, plat: Double, a: [Double], b: [Double], mLng: Double) -> Double {
        let ax = (a[0] - plng) * mLng, ay = (a[1] - plat) * metersPerDegLat
        let bx = (b[0] - plng) * mLng, by = (b[1] - plat) * metersPerDegLat
        let dx = bx - ax, dy = by - ay
        if dx == 0, dy == 0 { return (ax * ax + ay * ay).squareRoot() }
        var t = -(ax * dx + ay * dy) / (dx * dx + dy * dy)
        t = Swift.max(0, Swift.min(1, t))
        let cx = ax + t * dx, cy = ay + t * dy
        return (cx * cx + cy * cy).squareRoot()
    }

    /// ray-casting point-in-polygon test. x is lng, y is lat to match the
    /// stored [lng, lat] vertex order
    private static func contains(ring: Ring, x: Double, y: Double) -> Bool {
        var inside = false
        var j = ring.count - 1
        for i in 0..<ring.count {
            let xi = ring[i][0], yi = ring[i][1]
            let xj = ring[j][0], yj = ring[j][1]
            if (yi > y) != (yj > y), x < (xj - xi) * (y - yi) / (yj - yi) + xi {
                inside.toggle()
            }
            j = i
        }
        return inside
    }
}
